src/Compilers/CSharp/Test/Semantic/Semantics/IndexAndRangeTests.cs

```diff
@@ -15,6 +15,331 @@ public class IndexAndRangeTests : CompilingTestBase
+                // on the Index and Range types.
+                );
+            comp.VerifyEmitDiagnostics(
+                // (20,5): error CS0656: Missing compiler required member 'System.Index.GetOffset'
+                //     {
+                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, @"{
+        _ = this[^0];
+        _ = this[0..];
+    }").WithArguments("System.Index", "GetOffset").WithLocation(20, 5));
+        }
+
+        [Theory]
+        [InlineData("Start")]
+        [InlineData("End")]
+        public void PatternIndexAndRangeNoStartAndEnd(string propertyName)
+        {
+            var src = @"
+namespace System
+{
+    public struct Index
+    {
+        public Index(int value, bool fromEnd) { }
+        public static implicit operator Index(int value) => default;
+        public int GetOffset(int length) => 0;
+    }
+    public struct Range
+    {
+        public Range(Index start, Index end) { }
+        public Index " + propertyName + @" => default;
+    }
+}
+class C
+{
+    public int Length => 0;
+    public int this[int i] => 0;
+    public int Slice(int i, int j) => 0;
+    void M()
+    {
+        _ = this[^0];
+        _ = this[0..];
+    }
+}";
+            var comp = CreateCompilation(src);
+
+            comp.VerifyDiagnostics(
+                // At binding time we don't look for all the necessary members
+                // on the Index and Range types.
+                );
+
+            comp.VerifyEmitDiagnostics(
+                // (22,5): error CS0656: Missing compiler required member 'System.Range.get_{propertyName}'
+                //     {
+                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, @"{
+        _ = this[^0];
+        _ = this[0..];
+    }").WithArguments("System.Range", "get_" + (propertyName == "Start" ? "End" : "Start")).WithLocation(22, 5));
+        }
+
+        [Fact]
+        public void PatternIndexAndRangeNoOptionalParams()
+        {
+            var comp = CreateCompilationWithIndexAndRange(@"
+class C
+{
+    public int Length => 0;
+    public int this[int i, int j = 0] => i;
+    public int Slice(int i, int j, int k = 0) => i;
+    public void M()
+    {
+        _ = this[^0];
+        _ = this[0..];
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (9,18): error CS1503: Argument 1: cannot convert from 'System.Index' to 'int'
+                //         _ = this[^0];
+                Diagnostic(ErrorCode.ERR_BadArgType, "^0").WithArguments("1", "System.Index", "int").WithLocation(9, 18),
+                // (10,18): error CS1503: Argument 1: cannot convert from 'System.Range' to 'int'
+                //         _ = this[0..];
+                Diagnostic(ErrorCode.ERR_BadArgType, "0..").WithArguments("1", "System.Range", "int").WithLocation(10, 18));
+        }
+
+        [Fact]
+        public void PatternIndexAndRangeUseOriginalDefition()
+        {
+            var comp = CreateCompilationWithIndexAndRange(@"
+struct S1<T>
```

Comment:
Got it. It's not allowed to use the pattern on a type which only fits based on its type arguments, and wouldn't fit with different type arguments.